identity_iota_core/src/rebased/client/full_client.rs

```diff
@@ -253,18 +259,41 @@ impl<S> IdentityClient<S> {
     Ok(budget)
   }
 
-  async fn get_coin_for_transaction(&self) -> Result<iota_sdk::rpc_types::Coin, Error> {
-    let coins = self
-      .coin_read_api()
-      .get_coins(self.sender_address(), None, None, None)
-      .await
-      .map_err(|err| Error::GasIssue(format!("could not get coins; {err}")))?;
+  async fn get_coin_for_transaction(&self) -> Result<Coin, Error> {
+    const LIMIT: usize = 10;
+    let mut cursor = None;
+
+    loop {
+      let coins = self
+        .coin_read_api()
+        .get_coins(self.sender_address(), None, cursor, Some(LIMIT))
+        .await?;
+
+      if coins.data.is_empty() {
+        return Err(Error::GasIssue(format!(
+          "no coins found for address {}",
+          self.sender_address()
+        )));
+      }
+
+      if let Some(coin) = coins.data.into_iter().max_by_key(|coin| coin.balance) {
+        if coin.balance >= MINIMUM_BALANCE {
+          return Ok(coin);
+        }
+      }
```

Comment:
See if you like this. It removes one level of nesting
```suggestion
      let Some(coin) = coins.data.into_iter().max_by_key(|coin| coin.balance) else {
        return Err(Error::GasIssue(format!(
          "no coins found for address {}",
          self.sender_address()
        )));
      };
      
      if coin.balance >= MINIMUM_BALANCE {
        return Ok(coin);
      }
```